Packages/com.unity.inputsystem/InputSystem/Actions/InputBindingResolver.cs

```diff
@@ -227,132 +218,140 @@ public unsafe void AddActionMap(InputActionMap map)
+                                currentCompositeAction = null;
+                                currentCompositeActionIndexInMap = InputActionState.kInvalidIndex;
+                            }
 
-                        // Look up controls.
-                        //
-                        // NOTE: We continuously add controls here to `resolvedControls`. Once we've completed our
-                        //       pass over the bindings in the map, `resolvedControls` will have all the controls for
-                        //       the current map.
-                        var firstControlIndex = memory.controlCount + resolvedControls.Count;
-                        var numControls = 0;
-                        if (devicesForThisMap != null)
-                        {
-                            // Search in devices for only this map.
-                            var list = devicesForThisMap.Value;
-                            for (var i = 0; i < list.Count; ++i)
+                            // Look up controls.
+                            //
+                            // NOTE: We continuously add controls here to `resolvedControls`. Once we've completed our
+                            //       pass over the bindings in the map, `resolvedControls` will have all the controls for
+                            //       the current map.
+                            firstControlIndex = memory.controlCount + resolvedControls.Count;
+                            if (devicesForThisMap != null)
                             {
-                                var device = list[i];
-                                if (!device.added)
-                                    continue; // Skip devices that have been removed.
-                                numControls += InputControlPath.TryFindControls(device, path, 0, ref resolvedControls);
+                                // Search in devices for only this map.
+                                var list = devicesForThisMap.Value;
+                                for (var i = 0; i < list.Count; ++i)
+                                {
+                                    var device = list[i];
+                                    if (!device.added)
+                                        continue; // Skip devices that have been removed.
+                                    numControls += InputControlPath.TryFindControls(device, path, 0, ref resolvedControls);
+                                }
+                            }
+                            else
+                            {
+                                // Search globally.
+                                numControls = InputSystem.FindControls(path, ref resolvedControls);
                             }
-                        }
-                        else
-                        {
-                            // Search globally.
-                            numControls = InputSystem.FindControls(path, ref resolvedControls);
```

Comment:
The diff above looks scary. But the actual change is pretty straightforward.

Before:

```
1. Default-initialize *part* of the binding state.
2. Check whether we should skip the binding. If so, early out.
3. Resolve action. If action says we should skip the binding, early out.
4. Actually initialize the binding.
```

This left bindings that were skipped with a partially initialized state. In the revised versions, bindings are never just skipped.

Now:

```
1. Set up default values for everything that we will write into the binding state.
2. Resolve the binding's action (if any).
3. Now check whether we need to resolve the binding. If yes, then update all the default values above with initialized values.
4. Write the final values into the binding.
```

This initializes the binding state fully and also makes sure that *every* binding, regardless of whether it is used or not, ends up with a correct reference to the action it is associated with. This fixes a number of issues with binding data getting lost through various operations.